apis/analysis/v1alpha1/workload.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// AnalysisTargetType defines the type of analysis target
+type AnalysisTargetType string
+
+const (
+	// AnalysisTargetWorkload defines the k8s workload type
+	AnalysisTargetWorkload AnalysisTargetType = "workload"
+	// AnalysisTargetPodSelector defines the pod selector type
+	AnalysisTargetPodSelector AnalysisTargetType = "podSelector"
+)
+
+// AnalysisTarget defines the target of analysis, which can be a k8s workload or a series of pods
+type AnalysisTarget struct {
+	// Type indicates the type of target
+	Type AnalysisTargetType `json:"type"`
+	// Workload indicates the target is a k8s workload, which is effective when Type is "workload"
+	Workload *WorkloadRef `json:"workload,omitempty"`
+	// PodSelector indicates the target is a series of pods, which is effective when Type is "podSelector"
+	PodSelector *PodSelectorRef `json:"podSelector,omitempty"`
+}
+
+// WorkloadRef defines the reference of a k8s workload
+type WorkloadRef struct {
+	// Kind of the referent; More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds"
+	Kind string `json:"kind"`
+	// Name of the referent; More info: http://kubernetes.io/docs/user-guide/identifiers#names
+	Name string `json:"name"`
+	// API version of the referent
+	APIVersion string `json:"apiVersion,omitempty"`
+	// Hierarchy indicates the hierarchy of the target for profiling
+	Hierarchy ProfileHierarchy `json:"hierarchy,omitempty"`
```

Comment:
The `WorkloadRef` makes sense, but the field `Hierarchy` doesn't look connected to the workload reference, is the definition really appropriate here? I noticed that `PodSelectorRef` has the same situation, so `Hierarchy` itself is a field that needs to be described independently?

Reply:
The `Hierarchy` field is related to `Workload`, which means the metric prediction is container or pod level , which is only effective for K8s workloads.
Metric Prediction can also work for other types of workloads beyond K8s such as a FaaS job. Although the workload definition is not is K8s, the metric is recorded as prometheus format.
```
function_cpu_usage{service="service-word-count", function="f-map-name", slice="slice-0"} 1.2
function_cpu_usage{service="service-word-count", function="f-map-name", slice="slice-1"} 1.3
function_cpu_usage{service="service-word-count", function="f-reduce-name", slice="all"} 2
```
It means a workload called(service-word-count), which is consisted of two jobs(map and reduce).
For resource recommendation scenario, we want to profile `service-word-count/f-map-name` and `service-word-count/f-reduce-name`.
This workload can be defined as AnalysisTargetPromethuesLabelGroup, and we should support `service/function` as the key for aggregation in `PromethuesLabelGroup`, which does not need `Hierarchy` field.